developer-docs/architecture/type-system.md

```diff
@@ -0,0 +1,201 @@
+or `Archive`s, respectively. In the latter case, the referenced file or URL is an opaque
+blob or a TAR, gzipped TAR, or ZIP archive, respectively.
+
+Each `Asset` or `Archive` also carries the SHA-256 hash of its contents. This hash can be
+used to uniquely identify the asset (e.g. for locally caching `Asset` or `Archive`
+contents).
+
+### Resource References
+
+A `ResourceReference` represents a reference to a resource . Although
+all that is necessary to uniquely identify a resource within the context of a stack is its
+URN, a `ResourceReference` also carries the resource's ID (if it is not a component)) and
+the version of the provider that manages the resource. If the contents of the referenced
+resource must be inspected, the reference must be resolved by invoking the `getResource`
+function of the engine's builtin provider. Note that this is only possible if there is a 
+connection to the engine's resource monitor, e.g. within the scope of a call to `Construct`.
+This implies that resource references may not be resolved within calls to other 
+provider methods. Therefore, configuration values, custom resources and provider functions
+should not rely on the ability to resolve resource references, and should instead treat
+resource references  as either their ID (if present) or URN. If the ID is present and
+empty, it should be treated as an [`Unknown`](#unknowns).
+
+## Object Types
+
+Object types are defined as mapping from property names to property types. Duplicate
+property names are not allowed, and each property name maps to a single type.
+
+## `Promise<T>`
+
+Like promises in other systems, a value of type `Promise<T>` represents the result of an
+asynchronous computation or its failure.
+
+## `Output<T>`
+
+Perhaps the most important type in the Pulumi type system is `Output<T>`. A value of type
+`Output<T>` represents a node in a Pulumi program graph, and behaves like a `Promise<T>`
+that carries additional metadata that describes the resources on which the value depends,
+whether the value is known or unknown, and whether or not the value is secret. Readers
+familiar with functional programming will recognize `Output<T>` as a monad.
+
+### Dependencies
+
+If an `Output<T>` value is the result of a resource operation--e.g. if it is an output
+property of some resource--it is said to _depend_ on that resource.
+
+If a value of type `Output<T>` depends on a resource `R`, the result of any computation
+that depends on its concrete value also depends on `R`.
+
+### Unknowns 
+
+An `Output<T>` may be unknown if it depends on the result of a resource operation that
+will not be run because it is part of a `pulumi preview`. Previews typically produce for
+properties with values that cannot be determined until the resource is actually created or
+updated.
+
+If a value of type `Output<T>` is unknown, any computation that depends on its concrete
+value must not run, and may produce an unknown `Output<T>`.
```

Comment:
Can such a depending value produce a known `Output<T>`?

Reply:
The wording here is a bit obtuse, and was intended to leave the door open for nested unknowns--something that we use in practice, but something that isn't directly addressed in the doc and that I was hoping to leave out of the abstract model. Nested unknowns--i.e. unknown values inside of otherwise known object/map/array outputs--arise pretty frequently when dealing with previews, where they may be produced by resource providers as part of previewing a create or update operation.

Given the definitions in the current doc, the value inside a single output is either fully known or fully unknown. Given those semantics, the only way to build a partially-knowable value is to use nested outputs. This is essentially the mirror of `inputShape`, but for outputs, so let's look at the `Array<string>` example from the output perspective: if we want to produce an `Output<Array<string>>` where each element may be known or unknown, we need to use a type like `Output<Union<Array<string>, Array<Output<string>>>>`, or simply `Output<Array<Output<string>>>`. Projecting these types is pretty raw and unwieldy, as they can requires a great deal of unwrapping as nesting depth increases.

Instead of projecting these types, the various SDKs tend to opt to project the simpler `Output<Array<string>>` type, and use an internal representation for the concrete value that includes distinguished unknown values. This approach allows the SDKs to allow e.g. lifted property and element access into partially-known composite values. For example, the Node SDK will allow the user to access an element of an `Output<[]string>` via a proxied index operator even if some elements of the array are unknown, though it will _not_ allow the user to access the entire value via `apply`.

So to answer your original question: at the model level, "may produce an unknown" should be "must produce an unknown". I'll change the wording and work on adding a brief section regarding nested `Ouptut<T>` values after the combinators.